sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -511,6 +511,14 @@ class Dataset[T] private[sql](
    */
   def isLocal: Boolean = logicalPlan.isInstanceOf[LocalRelation]
 
+  /**
+   * Returns true if the `Dataset` is empty.
+   *
+   * @group basic
+   * @since 2.4.0
+   */
+  def isEmpty: Boolean = rdd.isEmpty()
```

Comment:
Building a rdd is not cheap. The current impl does not perform well.

Reply:
How about this?
```
  def isEmpty: Boolean = withAction("isEmpty", groupBy().count().queryExecution) { plan =>
    plan.executeCollect().head.getLong(0) == 0
  }
```

Reply:
@gatorsmile, just simply running df.rdd.isEmpty in spark-shell was quite responsive even in tera byte sized tables.

Reply:
@goungoun maybe; it depends on the Dataset. I'd follow @maropu 's lead or close this.

Reply:
If it's worth doing, I'll take (if nobody does this...)

Reply:
`RDD#isEmpty` is pretty effective, it just checks if all the partitions are empty, without loading the data. The problem is how to build an RDD from `Dataset`, which minimize the cost of building the `Iterator`.

It seems `Dataset#rdd` is not good enough, e.g., if we have a `Filter` in the query, we may do a full scan(no column pruning) for the underlying files.

Doing a count is not perfect either. Ideally we can stop as soon as we see one record.

I'd suggest doing a `limit 1` first and then count.

Reply:
+1 for the limit

Reply:
In my understanding, the @maropu's suggestion is to implement isEmpty exactly the same way of count() method except the return type. 

```
def count(): Long = withAction("count", groupBy().count().queryExecution) { plan =>
   plan.executeCollect().head.getLong(0)
 }

def isEmpty: Boolean = withAction("isEmpty", groupBy().count().queryExecution) { plan =>
  plan.executeCollect().head.getLong(0) == 0
}
```
And @cloud-fan 's idea is adding limit(1) before doing count! Is it right?